libresin/libresin/core/sdf_tree/group_node.cpp

```diff
@@ -0,0 +1,222 @@
+  insert_leaves_up(node_ptr);
+}
+
+void GroupNode::remove_from_parent(std::unique_ptr<SDFTreeNode>& node_ptr) {
+  node_ptr->remove_from_parent();
+  node_ptr->transform().remove_from_parent();
+  remove_leaves_up(node_ptr);
+}
+
+SDFTreeNode& GroupNode::get_child(IdView<SDFTreeNodeId> node_id) const {
+  auto it = nodes_.find(node_id);
+  if (it == nodes_.end()) {
+    log_throw(SDFTreeNodeIsNotAChild(node_id.raw(), node_id_.raw()));
+  }
+
+  return *it->second.second;
+}
+
+void GroupNode::delete_child(IdView<SDFTreeNodeId> node_id) {
+  auto it = nodes_.find(node_id);
+  if (it == nodes_.end()) {
+    log_throw(SDFTreeNodeIsNotAChild(node_id.raw(), node_id_.raw()));
+  }
+  remove_from_parent(it->second.second);
+
+  nodes_order_.erase(it->second.first);
+  nodes_.erase(it);
+}
+
+bool GroupNode::child_has_neighbor_prev(IdView<SDFTreeNodeId> node_id) const {
+  auto map_it = nodes_.find(node_id);
+  if (map_it == nodes_.end()) {
+    log_throw(SDFTreeNodeIsNotAChild(node_id.raw(), node_id_.raw()));
+  }
+
+  return map_it->second.first != nodes_order_.begin();
+}
+
+bool GroupNode::child_has_neighbor_next(IdView<SDFTreeNodeId> node_id) const {
+  auto map_it = nodes_.find(node_id);
+  if (map_it == nodes_.end()) {
+    log_throw(SDFTreeNodeIsNotAChild(node_id.raw(), node_id_.raw()));
+  }
+
+  auto it = map_it->second.first;
+  return ++it != nodes_order_.end();
+}
+
+SDFTreeNode& GroupNode::child_neighbor_prev(IdView<SDFTreeNodeId> node_id) {
+  auto map_it = nodes_.find(node_id);
+  if (map_it == nodes_.end()) {
+    log_throw(SDFTreeNodeIsNotAChild(node_id.raw(), node_id_.raw()));
+  }
+
+  auto list_it = map_it->second.first;
+  if (list_it == nodes_order_.begin()) {
+    log_throw(SDFTreeNodeIsNotAChild(node_id.raw(), node_id_.raw()));
+  }
+
+  return *nodes_.find(*(++list_it))->second.second;
+}
+
+SDFTreeNode& GroupNode::child_neighbor_next(IdView<SDFTreeNodeId> node_id) {
+  auto map_it = nodes_.find(node_id);
+  if (map_it == nodes_.end()) {
+    log_throw(SDFTreeNodeIsNotAChild(node_id.raw(), node_id_.raw()));
+  }
+
+  auto list_it = map_it->second.first;
+  if (++list_it == nodes_order_.end()) {
+    log_throw(SDFTreeNodeIsNotAChild(node_id.raw(), node_id_.raw()));
+  }
+
+  return *nodes_.find(*list_it)->second.second;
+}
+
+std::unique_ptr<SDFTreeNode> GroupNode::detach_child(IdView<SDFTreeNodeId> node_id) {
+  auto map_it = nodes_.find(node_id);
+  if (map_it == nodes_.end()) {
+    log_throw(SDFTreeNodeIsNotAChild(node_id.raw(), node_id_.raw()));
+  }
+
+  nodes_order_.erase(map_it->second.first);
+  auto child_ptr = std::move(map_it->second.second);
+  nodes_.erase(map_it);
+
+  remove_from_parent(child_ptr);
+
+  return child_ptr;
+}
+
+void GroupNode::push_back_child(std::unique_ptr<SDFTreeNode> node_ptr) {
+  set_parent(node_ptr);
+  auto node_id = node_ptr->node_id();
+
+  nodes_order_.push_back(node_id);
+  nodes_.emplace(node_id, std::make_pair(std::prev(nodes_order_.end()), std::move(node_ptr)));
```

Comment:
Can it be 
```suggestion
  nodes_.emplace(node_id, std::make_pair(nodes_order_.rbegin(), std::move(node_ptr)));
```

Reply:
the problem is that std::make_pair struggles to convert the reverse iterator into the forward iterator, so I would have to do this explicitly, which requires more code